Prevent naming properties with reserved words (Fix #16846)

kriswilk opened the pull request:
**Since this fixes a prominent new feature in 1.0, I suggest that it be reviewed prior to the official release.**

This PR fixes #16846. It prevents the use of reserved words (eg. `T`, `K`, `rad`, ...) for properties. It affects both the new VarSet feature and the pre-existing "Add Property" dialog box.

Please note:

- Although I've used FreeCAD for some time, this is my first code contribution. Apologies for any misstep(s)!
- I compiled and tested this fix on my WIndows 11 PC using the MSVC toolchain.

I look forward to feedback!

hyarion commented:
Welcome and congratulations on creating your first PR! 🎉 

I've looked though it and it looks good to me.

While you're at it, could you check if the constants `e` and `pi` causes the same problems?
If it does, can it be fixed at the same time? (if it's easy)

kriswilk commented:
> While you're at it, could you check if the constants e and pi causes the same problems?

Good question, @hyarion. Those tokens were not handled by the checks that I added. Furthermore, I discovered that the Spreadsheet workbench also suffers from this oversight (although it does check for unit names).

I've updated my PR to add a new utility function `App::ExpressionParser::isTokenAConstant()` and amended my previous code to include a check using it, too. Additionally, I fixed the Spreadsheet alias validation routine.

luzpaz commented:
Congrats on your first FreeCAD PR :tada:

chennes commented:
I'd like to add a unit test for this new piece of code, so I'm assigning it to myself.

kriswilk commented:
> I'd like to add a unit test for this new piece of code, so I'm assigning it to myself.

Thanks, I appreciate the care. I noticed that you removed "needs backport", which I understand if you're being strict about adding anything new before 1.0.

However in #16846 noted that users will quickly run into the issue if they use letters for VarSet properties (eg. sensible parameter names like `H`, `W`, `T`, `N`, ...). Worse, the problem presents very cryptically and cannot be fixed by renaming the property since there's no way to do so yet.

Since this is a prominent/useful new feature in 1.0, I suggest that you consider working this into 1.0 (properly tested, of course).

Thanks again for your attention!

chennes commented:
@kriswilk can you please take a look at the test I've added and confirm that it meets your expectations for the new method?

kriswilk commented:
@chennes Yes, your test looks comprehensive. 👍

pieterhijma commented:
@kriswilk, thanks for submitting this PR. Although I was aware of the issue, I wasn't notified about this PR. Otherwise, I would have gladly reviewed it as well. Looks good!
